man/cmds/fscmd.man
' $Header: /sprite/src/cmds/fscmd/RCS/fscmd.man,v 1.2 88/12/30 09:28:33 ouster Exp Locker: shirriff $ SPRITE (Berkeley)
.so \*(]ltmac.sprite
.HS FSCMD cmds
.BS
.SH NAME
fscmd \- Set various parameters in the filesystem
.SH SYNOPSIS
.nf
\fBfscmd\fR \-[\fBmM\fR] \fIsize\fR
\fBfscmd\fR \-[\fBfslLNCv\fR] \fR
\fBfscmd\fR \-[\fBwWxXyz\fR] \fR
\fBfscmd\fR \-[\fBBOtDR\fR] \fR
\fBfscmd\fR \-\fBgeneric\fR \fIarg\fR
.fi
.BE
.SH INTRODUCTION
This command is used to set various kernel variables that control
the filesystem cache, tracing and debugging output, and other
features of the system.
It is implemented using the Fs_Command system call.
.SH CACHE SIZE
The minimum and maximum number of filesystem blocks (4 Kbytes each) in the
main memory cache is set with the -m (minimum) and -M (maximum)
arguments.  Follow either of these with the number of 4-Kbyte blocks, ie.
-M 512 means a max of 2 Megabytes in the cache.
.SH CACHE CONTROL
.IP "\fB\-f"
The entire contents of the cache gets written-back and then invalidated by
specifying \fB\-f\fP.
.IP "\fB\-s\fR [\fB0\fR|\fB1\fR]"
The background process that writes back dirty data from the cache
can be
disabled
by specifying the
``\fB\-s 0\fP''
flag.  This should be done when patching a mounted filesystem
or else bad internal structures will get pushed to disk.
Enable the sync daemon with ``\fB\-s 1\fP''.
.IP "\fB\-l\fR [\fB0\fR|\fB1\fR]"
Biasing against ``large'' files in the cache is enabled with ``\fB\-l 1\fR''
and turned off with ``\fB\-l 0\fR''.
.IP "\fB\-L \fIsize\fR"
This sets what portion of the maximum cache size a file must occupy before
it is considered
a large file.  For example a value of 8 would indicate that if a file is larger
than 1/8 of the maximum cache size then it is considered a large file.
.IP "\fB\-C\fR [\fB0\fR|\fB1\fR]"
Caching on diskless nodes can be disabled with ``\fB\-C 0\fR'',
and turned back on with ``\fB\-C 1\fR''.
This has to be done on a file server to affect its clients; use of the -C
flag on clients will have no effect.
.IP "\fB\-N\fR [\fB0\fR|\fB1\fR]"
The name cache on a diskfull node is disabled with
``\fB\-N 0\fR'' and enabled with ``\fB\-N 1\fR''.
.IP "\fB\-v\fR [\fB0\fR|\fB1\fR]"
The Virtual Memory cache of recently used program segments is
disabled with ``\fB\-v 1\fR'' (yes, 1), and enabled with ``\fB\-v 0\fR''.
.IP "\fB\-b \fIN\fR"
Set the maximum number of block cleaner processes.  This controls the
amount of overlap during cache writebacks.  3 processes provides the most
bandwidth, but it does load the server.
.IP "\fB\-r \fIN\fR"
Set the read ahead distance (in 4K blocks).  When a read on the cache
is done, the transfer of N more blocks is initiated.
.SH "CACHE WRITE BACK POLICY"
Normally, dirty blocks in the file system cache are aged for at least
30 seconds before being written back.  This gives the best performance,
although alternate strategies can be tried out with the following flags.
There are two write-back policies involved: the client's and the server's.
Each of the flags below only affects the policy on the client or server
on which it is invoked.  For example the client could be using a
write-through policy and the server a 30-second-delay policy.  This would
mean that client writes would not complete until the blocks have been written
across the network into the server's cache, but the blocks would not be
written back to disk until at least 30 seconds later.
.IP "\fB\-S \fIseconds\fR"
Set the cache write-back interval in seconds. 
.IP "\fB\-w\fR [\fB0\fR|\fB1\fR]"
Write-through caching is enabled with ``\fB\-w 1\fR'' and
disabled with ``\fB\-w 0\fR''.  With write-through,
the write system call does not complete until
the cache block is written back. 
.IP "\fB\-W [\fB0\fR|\fB1\fR]"
Write-back-on-close caching is enabled with
``\fB\-W 1\fR'' and disabled with ``\fB\-W 0\fR''.
With write-back-on-close the close system call does not complete
until all of the file's dirty blocks are written back.  
.IP "\fB\-x\fR [\fB0\fR|\fB1\fR]"
Write-back of temporary files can be set to delay as long as possible
with ``\fB\-x 1\fR''.  This is turned off with ``\fB\-x 0\fR''.  Normally temporary files are
files that are in the /tmp directory.
.IP "\fB\-X \fIdirNumber\fR"
This identifies the temporary directory to the file system.
\fIdirNumber\fP should be the inode number of /tmp.
Unfortunately, only one temporary directory is understood.
.IP "\fB\-y\fR [\fB0\fR|\fB1\fR]
Write-back-as-soon-as-possible caching is enabled with ``\fB\-y 1\fR''
and turned off with ``\fB\-y 0\fR''.
With write-back-ASAP, the write system call completes immediately
and the cache block is scheduled to be written out as-soon-as-possible.
.IP "\fB\-z\fR [\fB0\fR|\fB1\fR]"
Write-back-on-last-dirty-block caching is enabled with ``\fB\-z 1\fR''
and turned off with ``\fB\-z 0\fR''.
This policy only works on a server.  It 
means the server writes back its blocks for a file
after getting the last dirty block for the file from a client.  For example,
this policy combined with a write-back-on-close client policy would mean
that after a client closes a file, all of the files blocks are guaranteed
to be on the server's disk.
.SH DISK LAYOUT
.IP "\fB\-B \fIskew\fR"
Set the block skew for disk allocation.  Not supported in kernel yet.
.SH TRACING/DEBUG FLAGS
There are several tracing/debugging flags that can have their
value set via \fBfscmd\fR.  Each flag expects a following integer
on the command line that will specify the new value of
the flag or trace level variable.  \fBFscmd\fR prints out the
old value of the flag as it changes it.
Some of the flags are Booleans that are
set with values of 1 for TRUE, and 0 for FALSE.
.IP "\fB\-O\fR"
The fsStats structure is zeroed out.  This information is displayed
with the fsStat program.
.IP "\fB\-t\fR [\fB0\fR|\fB1\fR]"
Filesystem tracing is enabled with ``\fB\-t 1\fR''.  This dumps internal events
into a circular trace record.  The trace gets printed via a
special console key sequence, ``\fBL1-f\fR''.
.IP "\fB\-D\fR [\fB0\fR|\fB1\fR]"
Cache consistency debug print statements are turned on with ``\fB\-D 1\fR'',
and turned off with ``\fB\-D 0\fR''.
.IP "\fB\-R\fR [\fB0\fR|\fB1\fR]"
Read ahead tracing is enabled with ``\fB\-R 1\fR'' and turned off with ``\fB\-R 0\fR''.
.IP "\fB\-generic\fR \fIarg\fR"
A generic hook for temporary file system operations.  The result depends on
arg.
.SH "SEE ALSO"
fsStat, prefix, rpcCmd
.SH FILES
.nf
/local/diskcmds		- Server local startup file
/bootcmds		- Network wide startup file
.SH KEYWORDS
debugging, trace, cache, write-back


src/machserver/src/cmds/fscmd/fscmd.c
/* 
 * fscmd.c --
 *
 *	Command-line interface to the Fs_Command system call.
 *	This is used to set various filesystem parameters, ie.
 *	min/max cache size, cache write-back, debug printing.
 *
 * Copyright (C) 1986 Regents of the University of California
 * All rights reserved.
 */

#ifndef lint
static char rcsid[] = "$Header: /sprite/src/cmds/fscmd/RCS/fscmd.c,v 1.20 90/02/16 11:26:37 brent Exp $ SPRITE (Berkeley)";
#endif not lint

#include "sprite.h"
#include "status.h"
#include "option.h"
#include "fs.h"
#include "fsCmd.h"
#include "stdio.h"
/*
 * Command line options.
 */

int minBlocks = -1;
int maxBlocks = -1;
int	cacheWriteBack = -1;
Boolean	flushCache = FALSE;
int	fsTracing = -1;
int	cacheDebug = -1;
int	pdevDebug = -1;
int	migDebug = -1;
int	nameCaching = -1;
int	clientCaching = -1;
int	contextSwitches = -1;
int	noStickySegs = -1;
int	numCleanerProcs = -1;
int	readAhead = -1;
int	readAheadTracing = -1;
int	writeThrough = -1;
int	writeBackOnClose = -1;
int	delayTmpFiles = -1;
int	tmpDirNum = -1;
int	writeBackASAP = -1;
int	wbOnLastDirtyBlock = -1;
int	writeBackInterval = -1;
int	blockSkew = -1;
int	largeFileMode = -1;
int	maxFilePortion = -1;
int	deleteHistograms = -1;
Boolean zeroStats = FALSE;
char    *updateSummary = NULL;

Option optionArray[] = {
    {OPT_INT, "B", (Address) &blockSkew, "\tSet block gap for allocation."},
    {OPT_INT, "C", (Address) &clientCaching, "\tSet client caching flag."},
    {OPT_INT, "D", (Address) &cacheDebug, "\tSet cache debug flag."},
    {OPT_INT, "H", (Address)&deleteHistograms, "\tSet delete histograms flag."},
    {OPT_INT, "L", (Address)&maxFilePortion, "\tSet large file portion."},
    {OPT_INT, "M", (Address) &maxBlocks, "\tMaximum number of blocks allowed in file system cache."},
    {OPT_INT, "N", (Address) &nameCaching, "\tSet name caching flag."},
    {OPT_TRUE, "O", (Address)&zeroStats, "\tZero fsStat counters."},
    {OPT_INT, "R", (Address) &readAheadTracing, "\tSet read ahead tracing flag."},
    {OPT_INT, "S", (Address) &writeBackInterval, "\tSet cache write back interval."},
    {OPT_INT, "W", (Address) &writeBackOnClose, "\tSet write-back-on-close flag"},
    {OPT_INT, "X", (Address) &tmpDirNum, "\tSet tmp directory number."},
    {OPT_INT, "Z", (Address) &contextSwitches, "\tDo context switches."},
    {OPT_INT, "b", (Address) &numCleanerProcs, "\tSet the maximum number of block cleaner processes."},
    {OPT_TRUE, "f", (Address) &flushCache, "\tFlush and invalidate cache."},
    {OPT_INT, "g", (Address) &migDebug, "\tSet migration debug flag."},
    {OPT_INT, "l", (Address)&largeFileMode, "\tSet large file mode for cache."},
    {OPT_INT, "m", (Address) &minBlocks, "\tMinimum number of blocks allowed in file system cache."},
    {OPT_INT, "p", (Address) &pdevDebug, "\tSet pseudo-device debug flag."},
    {OPT_INT, "r", (Address) &readAhead, "\tSet number of blocks of read ahead."},
    {OPT_INT, "s", (Address) &cacheWriteBack, "\tSet cache write back flag."},
    {OPT_INT, "t", (Address) &fsTracing, "\tSet fs tracing value."},
    {OPT_STRING, "u", (Address)&updateSummary, "\tUpdate summary info from disk."},
    {OPT_INT, "v", (Address) &noStickySegs, "\tSet no sticky segments flag."},
    {OPT_INT, "w", (Address) &writeThrough, "\tSet write through flag."},
    {OPT_INT, "x", (Address) &delayTmpFiles, "\tSet delay tmp files flag."},
    {OPT_INT, "y", (Address) &writeBackASAP, "\tSet write back ASAP flag."},
    {OPT_INT, "z", (Address) &wbOnLastDirtyBlock, "\tSet write back on last dirty block flag."},
};
int numOptions = sizeof(optionArray) / sizeof(Option);


/*
 *----------------------------------------------------------------------
 *
 * main --
 *
 *	Collects arguments and branch to the code for the fs command.
 *
 * Results:
 *	None.
 *
 * Side effects:
 *	Calls Fs_Command...
 *
 *----------------------------------------------------------------------
 */
main(argc, argv)
    int argc;
    char *argv[];
{
    register ReturnStatus status = SUCCESS;	/* status of system calls */

    argc = Opt_Parse(argc, argv, optionArray, numOptions);

    if (minBlocks != -1) {
	(void) Fs_Command(FS_RAISE_MIN_CACHE_SIZE, sizeof(int), &minBlocks);
    }
    if (maxBlocks != -1) {
	(void) Fs_Command(FS_LOWER_MAX_CACHE_SIZE, sizeof(int), &maxBlocks);
    }
    if (flushCache) {
	/*
	 * Flush the filesystem cache.
	 */
	int numLockedBlocks;
	status = Fs_Command(FS_EMPTY_CACHE, sizeof(int),
					    (Address) &numLockedBlocks);
	if (numLockedBlocks > 0) {
	    fprintf(stderr, "%d locked blocks left\n",
				      numLockedBlocks);
	}
    }
    if (zeroStats) {
	/*
	 * Zero out the file system statistics.
	 */
	(void) Fs_Command(FS_ZERO_STATS, 0, (Address) NULL);
    }
    if (updateSummary != NULL) {
	/*
	 * Reread summary information from disk and update in-memory copy.
	 */
	status = Fs_Command(FS_REREAD_SUMMARY_INFO, strlen(updateSummary)+1,
			updateSummary);
    }
    if (writeBackInterval != -1) {
	int	newInterval;

	newInterval = writeBackInterval;
	Fs_Command(FS_SET_WRITE_BACK_INTERVAL, sizeof(int),
		   &writeBackInterval);
	printf("Cache write back interval was %d, now %d\n",
		writeBackInterval, newInterval);
    }
    /*
     * Set various file system flags.  The Fs_Command system call returns
     * the old value of the flag in place of the value passed in.
     */
    if (cacheWriteBack != -1) {
	/*
	 * Set the cache write-back flag.
	 */
	status = SetFlag(FS_DISABLE_FLUSH, cacheWriteBack, "Cache write-back");
    }
    if (fsTracing != -1) {
	SetFlag(FS_SET_TRACING, fsTracing, "Filesystem tracing");
    }
    if (cacheDebug != -1) {
	status = SetFlag(FS_SET_CACHE_DEBUG, cacheDebug, "Cache debug prints");
    }
    if (migDebug != -1) {
	status = SetFlag(FS_SET_MIG_DEBUG, migDebug, "Migration debug prints");
    }
    if (pdevDebug != -1) {
	status = SetFlag(FS_SET_PDEV_DEBUG, pdevDebug,
			 "pseudo-device debug prints");
    }
    if (nameCaching != -1) {
	status = SetFlag(FS_SET_NAME_CACHING, nameCaching, "Name caching");
    }
    if (clientCaching != -1) {
	status = SetFlag(FS_SET_CLIENT_CACHING, clientCaching, "Client caching");
    }
    if (noStickySegs != -1) {
	status = SetFlag(FS_SET_NO_STICKY_SEGS, noStickySegs, 
			 "No sticky segments");
    }

    if (contextSwitches != -1) {
	status = Fs_Command(FS_TEST_CS, 4, (Address) &contextSwitches);
    }
    if (numCleanerProcs != -1) {
	status = SetInt(FS_SET_CLEANER_PROCS, numCleanerProcs,
			"Num block cleaner procs");
    }
    if (readAhead != -1){
	status = SetInt(FS_SET_READ_AHEAD, readAhead,
			"Num blocks of read ahead");
    }
    if (readAheadTracing != -1) {
	status = SetFlag(FS_SET_RA_TRACING, readAheadTracing, 
			"Read ahead tracing");
    }
    if (writeThrough != -1) {
	status = SetFlag(FS_SET_WRITE_THROUGH, writeThrough, "Write through");
    }
    if (writeBackOnClose != -1) {
	status = SetFlag(FS_SET_WRITE_BACK_ON_CLOSE, writeBackOnClose,
		"Write-back-on-close");
    }
    if (delayTmpFiles != -1) {
	status = SetFlag(FS_SET_DELAY_TMP_FILES, delayTmpFiles,
			 "Delay tmp files");
    }
    if (tmpDirNum != -1) {
	status = Fs_Command(FS_SET_TMP_DIR_NUM, sizeof(int), &tmpDirNum);
    }
    if (writeBackASAP != -1) {
	status = SetFlag(FS_SET_WRITE_BACK_ASAP, writeBackASAP,
			 "Write back ASAP");
    }
    if (wbOnLastDirtyBlock != -1) {
	status = SetFlag(FS_SET_WB_ON_LAST_DIRTY_BLOCK, wbOnLastDirtyBlock,
			 "Write back on last dirty block");
    }
    if (blockSkew != -1) {
	status = SetInt(FS_SET_BLOCK_SKEW, blockSkew,
			 "Block gap for allocation");
    }
    if (largeFileMode != -1) {
	status = SetFlag(FS_SET_LARGE_FILE_MODE, largeFileMode,
			 "Large file mode for cache");
    }
    if (maxFilePortion != -1) {
	status = SetFlag(FS_SET_MAX_FILE_PORTION, maxFilePortion,
			 "Maximum large file portion for cache");
    }
    if (deleteHistograms != -1) {
	status = SetFlag(FS_SET_DELETE_HISTOGRAMS, deleteHistograms,
			 "Keep histograms of size/age distributions on delete");
    }
    exit(status);
}

ReturnStatus
SetInt(command, value, comment)
    int command;		/* Argument to Fs_Command */
    int value;			/* Value for int */
    char *comment;		/* For Io_Print */
{
    register int newValue;
    register ReturnStatus status;

    newValue = value;
    status = Fs_Command(command, sizeof(int), (Address) &value);
    if (status != SUCCESS) {
	printf("%s failed <%x>\n", comment, status);
    } else {
	printf("%s %d, was %d\n", comment, newValue, value);
    }
    return(status);
}

ReturnStatus
SetFlag(command, value, comment)
    int command;		/* Argument to Fs_Command */
    int value;			/* Value for flag */
    char *comment;		/* For Io_Print */
{
    register int newValue;
    register ReturnStatus status;

    newValue = value;
    status = Fs_Command(command, sizeof(int), (Address) &value);
    if (status != SUCCESS) {
	printf("%s failed <%x>\n", comment, status);
    } else {
	printf("%s %s, was %s\n", comment,
		     newValue ? "on" : "off",
		     value ? "on" : "off");
    }
    return(status);
}
